rust/arrow/src/ipc/writer.rs

```diff
@@ -0,0 +1,461 @@
+//!
+//! The `FileWriter` and `StreamWriter` have similar interfaces,
+//! however the `FileWriter` expects a reader that supports `Seek`ing
+
+use std::io::{BufWriter, Write};
+
+use flatbuffers::FlatBufferBuilder;
+
+use crate::array::ArrayDataRef;
+use crate::buffer::{Buffer, MutableBuffer};
+use crate::datatypes::*;
+use crate::error::{ArrowError, Result};
+use crate::ipc;
+use crate::record_batch::RecordBatch;
+
+pub struct FileWriter<W: Write> {
+    /// The object to write to
+    writer: BufWriter<W>,
+    /// A reference to the schema, used in validating record batches
+    schema: Schema,
+    /// The number of bytes written for the header (up to schema)
+    header_bytes: usize,
+    /// The number of bytes between each block of bytes, as an offset for random access
+    block_offsets: usize,
+    /// Dictionary blocks that will be written as part of the IPC footer
+    dictionary_blocks: Vec<ipc::Block>,
+    /// Record blocks that will be written as part of the IPC footer
+    record_blocks: Vec<ipc::Block>,
+    /// Whether the writer footer has been written, and the writer is finished
+    finished: bool,
+}
+
+impl<W: Write> FileWriter<W> {
+    /// Try create a new writer, with the schema written as part of the header
+    pub fn try_new(writer: W, schema: &Schema) -> Result<Self> {
+        let mut writer = BufWriter::new(writer);
+        // write magic to header
+        writer.write(&super::ARROW_MAGIC[..])?;
+        // create an 8-byte boudnary after the header
+        writer.write(&[0, 0])?;
+        // write the schema, set the written bytes to the schema + header
+        let written = write_schema(&mut writer, schema)? + 8;
+        Ok(Self {
+            writer,
+            schema: schema.clone(),
+            header_bytes: written,
+            block_offsets: written,
+            dictionary_blocks: vec![],
+            record_blocks: vec![],
+            finished: false,
+        })
+    }
+
+    /// Write a record batch to the file
+    pub fn write(&mut self, batch: &RecordBatch) -> Result<()> {
+        if self.finished {
+            return Err(ArrowError::IoError(
+                "Cannot write record batch to file writer as it is closed".to_string(),
+            ));
+        }
+        let (meta, data) = write_record_batch(&mut self.writer, batch)?;
+        // add a record block for the footer
+        self.record_blocks.push(ipc::Block::new(
+            self.block_offsets as i64,
+            (meta as i32) + 4,
+            data as i64,
+        ));
+        self.block_offsets += meta + data;
+        Ok(())
+    }
+
+    /// write footer and closing tag, then mark the writer as done
+    pub fn finish(&mut self) -> Result<()> {
+        let mut fbb = FlatBufferBuilder::new();
+        let dictionaries = fbb.create_vector(&self.dictionary_blocks);
+        let record_batches = fbb.create_vector(&self.record_blocks);
+        // TODO: this is duplicated as we otherwise mutably borrow twice
```

Comment:
I need help here, I couldn't find a way of reusing the functions in `crate::ipc::convert`

Reply:
I pulled this change down and played with it; I'm pretty excited to get the IPC writer for Rust :).  I've run into similar challenges before with flatbuffers, and sometimes there's no good way of working around it. 

Would it be worth converting schema_to_fb_offset to a macro maybe?

Reply:
I can try the macro approach if I can't find another workaround, I could do this when working on the `StreamWriter` given that it'll share most code with the `FileWriter`.